README.md
![logo](https://raw.githubusercontent.com/SyncDB/DateParser/master/Resources/logo-v1.png)

[![Version](https://img.shields.io/cocoapods/v/DateParser.svg?style=flat)](https://cocoapods.org/pods/DateParser)
[![Carthage compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/SyncDB/DateParser)
![platforms](https://img.shields.io/badge/platforms-iOS%20%7C%20OS%20X%20%7C%20watchOS%20%7C%20tvOS%20-lightgrey.svg)
[![License](https://img.shields.io/cocoapods/l/DateParser.svg?style=flat)](https://cocoapods.org/pods/DATAStack)

Simple ISO 8601 and Unix timestamp Swift date parser.

## Usage

```swift
// ISO 8601
let isoDate = NSDate(fromDateString:"2015-06-23T14:40:08.000+02:00")

// Unix Timestamp
let timestampDate = NSDate(fromDateString:"1441843200")

// Returns DateType.ISO8601  
let isoDateType = "2014-01-02T00:00:00.007450+00:00".dateType()

// Returns DateType.UnixTimestamp
let timestampDateType = "1441843200000000".dateType()
```

## Supported Formats

```
2014-01-02
2016-01-09T00:00:00
2014-03-30T09:13:00Z
2016-01-09T00:00:00.00
2015-06-23T19:04:19.911Z
2014-01-01T00:00:00+00:00
2015-09-10T00:00:00.184968Z
2015-09-10T00:00:00.116+0000
2015-06-23T14:40:08.000+02:00
2014-01-02T00:00:00.000000+00:00
```

## Installation

**DateParser** is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your Podfile:

```ruby
pod 'DateParser'
```

**DateParser** is also available through [Carthage](https://github.com/Carthage/Carthage). To install
it, simply add the following line to your Cartfile:

```ruby
github "SyncDB/DateParser"
```

## License

**DateParser** is available under the MIT license. See the LICENSE file for more info.

## Author

Elvis Nuñez, [@3lvis](https://twitter.com/3lvis)


CHANGELOG.md
Check https://github.com/SyncDB/DateParser/releases for more information.


Tests/ObjCTests.m
#import "Tests-Swift.h"

@interface ObjCTests : XCTestCase

@end

@implementation ObjCTests

- (void)setUp {
    [super setUp];
}

- (void)tearDown {
    [super tearDown];
}

- (void)testDateString {
    
}

@end


Tests/Tests.swift
import XCTest

class DateTests: XCTestCase {
    func testDateA() {
        let date = Date.dateWithDayString("2016-01-09")
        let resultDate = Date(dateString: "2016-01-09T00:00:00")
        XCTAssertNotNil(resultDate)
        XCTAssertEqual(date, resultDate)
    }

    func testDateB() {
        let date = Date.dateWithHourAndTimeZoneString("2014-03-30T09:13:00.000Z")
        let resultDate = Date(dateString: "2014-03-30T09:13:00Z")
        XCTAssertNotNil(resultDate)
        XCTAssertEqual(date, resultDate)
    }

    func testDateC() {
        let date = Date.dateWithDayString("2016-01-09")
        let resultDate = Date(dateString: "2016-01-09T00:00:00.00")
        XCTAssertNotNil(resultDate)
        XCTAssertEqual(date, resultDate)
    }

    func testDateD() {
        let date = Date.dateWithHourAndTimeZoneString("2015-06-23T19:04:19.911Z")
        let resultDate = Date(dateString: "2015-06-23T19:04:19.911Z")
        XCTAssertNotNil(resultDate)
        XCTAssertEqual(date, resultDate)
    }

    func testDateE() {
        let date = Date.dateWithDayString("2014-01-01")
        let resultDate = Date(dateString: "2014-01-01T00:00:00+00:00")

        XCTAssertNotNil(resultDate)
        XCTAssertEqual(date, resultDate)
    }

    func testDateF() {
        let date = Date.dateWithDayString("2014-01-02")
        let resultDate = Date(dateString: "2014-01-02")

        XCTAssertNotNil(resultDate)
        XCTAssertEqual(date, resultDate)
    }

    func testDateG() {
        let date = Date.dateWithDayString("2015-09-10")
        let resultDate = Date(dateString: "2015-09-10T00:00:00.000000Z")

        XCTAssertNotNil(resultDate)
        XCTAssertEqual(date, resultDate)
    }
    
    func testDateH() {
        let date = Date.dateWithDayString("2015-09-10")
        let resultDate = Date(dateString: "2015-09-10T00:00:00.000+0000")

        XCTAssertNotNil(resultDate)
        XCTAssertEqual(date, resultDate)
    }

    func testDateI() {
        let date = Date.dateWithHourAndTimeZoneString("2015-06-23T12:40:08.000")
        let resultDate = Date(dateString: "2015-06-23T14:40:08.000+02:00")

        XCTAssertNotNil(resultDate)
        XCTAssertEqual(date.timeIntervalSince1970, resultDate!.timeIntervalSince1970)
    }

    func testDateJ() {
        let date = Date.dateWithDayString("2014-01-02")
        let resultDate = Date(dateString: "2014-01-02T00:00:00.000000+00:00")

        XCTAssertNotNil(resultDate)
        XCTAssertEqual(date, resultDate)
    }

    /*
    // No longer testable since now it will use fatalError instead of returning nil
    func testDateK() {
        let resultDate = Date(dateString: "2014-01-02T00:monsterofthelakeI'mhere00:00.007450+00:00")
        XCTAssertNil(resultDate)        
    }
    */

}

class TimestampDateTests: XCTestCase {
    func testTimestampA() {
        let date = Date.dateWithDayString("2015-09-10")
        let resultDate = Date(dateString: "1441843200")

        XCTAssertNotNil(resultDate)
        XCTAssertEqual(date, resultDate)
    }

    func testTimestampB() {
        let date = Date.dateWithDayString("2015-09-10")
        let resultDate = Date(dateString: "1441843200000000")

        XCTAssertNotNil(resultDate)
        XCTAssertEqual(date, resultDate)
    }

    func testTimestampC() {
        let date = Date.dateWithDayString("2015-09-10")
        let resultDate = Date(unixTimestampNumber: 1441843200)

        XCTAssertNotNil(resultDate)
        XCTAssertEqual(date, resultDate)
    }

    func testTimestampD() {
        let date = Date.dateWithDayString("2015-09-10")
        let resultDate = Date(unixTimestampNumber: NSNumber(value: 1441843200000000.0))

        XCTAssertNotNil(resultDate)
        XCTAssertEqual(date, resultDate)
    }

    func testNegativeTimestamps() {
        let date = Date.dateWithHourAndTimeZoneString("1963-01-01T02:42:00.000")
        let resultDate = Date(unixTimestampNumber: NSNumber(value: -2209150800000))

        XCTAssertNotNil(resultDate)
        XCTAssertEqual(date, resultDate)
    }
}

class OtherDateTests: XCTestCase {
    func testDateType() {
        self.measure {
            let isoDateType = "2014-01-02T00:00:00.007450+00:00".dateType()
            XCTAssertEqual(isoDateType, DateType.iso8601)

            let timestampDateType = "1441843200000000".dateType()
            XCTAssertEqual(timestampDateType, DateType.unixTimestamp)

            let negativeTimestampDateType = "-2209150800000".dateType()
            XCTAssertEqual(negativeTimestampDateType, DateType.unixTimestamp)
        }
    }
}

extension Date {
    static func dateWithDayString(_ dateString: String) -> Date {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        formatter.timeZone = TimeZone(identifier: "GMT")
        let date = formatter.date(from: dateString)!

        return date
    }

    static func dateWithHourAndTimeZoneString(_ dateString: String) -> Date {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"
        formatter.timeZone = TimeZone(identifier: "GMT")
        let date = formatter.date(from: dateString)!

        return date
    }

    func prettyPrint() {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"
        let string = formatter.string(from: self)
        print(string)
    }
}


Tests/Tests-Bridging-Header.h
#import "NSDate+Parser.h"

Package.swift
// Licensed under the **MIT** license
// Copyright (c) 2016 Elvis Nuñez
// Copyright (c) 2016 SyncDB
//
// Permission is hereby granted, free of charge, to any person obtaining
// a copy of this software and associated documentation files (the
// "Software"), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to
// permit persons to whom the Software is furnished to do so, subject to
// the following conditions:
//
// The above copyright notice and this permission notice shall be
// included in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

import PackageDescription

let package = Package(
  name: "DateParser"
)


Sources/DateParser.swift
import Foundation

struct DateParser {
    static var noTimestampFormat = "YYYY-MM-DD"
    static var timestamp = "T00:00:00+00:00"
}

public enum DateType {
    case iso8601, unixTimestamp
}

public extension Date {
    public init?(dateString: String) {
        let dateType = dateString.dateType()
        switch dateType {
        case .iso8601:
            self.init(iso8601String: dateString)
            break
        case .unixTimestamp:
            self.init(unixTimestampString: dateString)
            break
        }
    }

    public init?(iso8601String: String) {
        if (iso8601String as NSString) == NSNull() {
            return nil
        } else {
            var dateString = iso8601String
            if dateString.characters.count == DateParser.noTimestampFormat.characters.count {
                dateString.append(DateParser.timestamp)
            }

            if iso8601String.isEmpty {
                return nil
            }

            let originalLength = dateString.characters.count
            let originalString = dateString
            var currentString = ""
            var hasFullTimeZone = false
            var hasNormalizedTimeZone = false
            var hasCentiseconds = false
            var hasMiliseconds = false
            var hasMicroseconds = false

            // ----
            // In general lines, if a Z is found, then the Z is removed since all dates operate
            // in GMT as a base, unless they have timezone, and Z is the GMT indicator.
            //
            // If +00:00 or any number after + is found, then it means that the date has a timezone.
            // This means that `hasTimezone` will have to be set to true, and since all timezones go to
            // the end of the date, then they will be parsed at the end of the process and appended back
            // to the parsed date.
            //
            // If after the date theres `.` and a number `2014-03-30T09:13:00.XXX` the `XXX` is the milisecond
            // then `hasMiliseconds` will be set to true. The same goes for `XX` decisecond (hasCentiseconds set to true).
            // and microseconds `XXXXXX` (hasMicroseconds set yo true).
            //
            // If your date format is not supported, then you'll get "Signal Sigabrt". Just ask your format to be included.
            // ----

            switch originalLength {
            case 19:
                // Copy all the date
                // Current date: 2014-03-30T09:13:00
                // Will become:  2014-03-30T09:13:00
                // Unit test A
                currentString = originalString
            case 20:
                // Copy all the date excluding the Z.
                // Current date: 2014-03-30T09:13:00Z
                // Will become:  2014-03-30T09:13:00
                // Unit test B
                let timeZoneIndicatorIndex = originalString.index(originalString.endIndex, offsetBy: -1)
                if originalString[timeZoneIndicatorIndex] == "Z" {
                    currentString = originalString.substring(to: timeZoneIndicatorIndex)
                }
            case 22:
                // Copy all the date excluding the miliseconds.
                // Current date: 2016-01-09T00:00:00.00
                // Will become:  2016-01-09T00:00:00
                // Unit test C
                let evaluatedDate = "2016-01-09T00:00:00"
                let timeZoneColonIndex = originalString.index(originalString.startIndex, offsetBy: evaluatedDate.characters.count)
                if originalString[timeZoneColonIndex] == "." {
                    let targetDate = "2014-01-01T00:00:00"
                    let baseDateIndex = originalString.index(originalString.startIndex, offsetBy: targetDate.characters.count)
                    currentString = originalString.substring(to: baseDateIndex)
                    hasCentiseconds = true
                }
            case 24:
                // Copy all the date excluding the miliseconds and the Z.
                // Current date: 2014-03-30T09:13:00.000Z
                // Will become:  2014-03-30T09:13:00
                // Unit test D
                let gmtIndicatorIndex = originalString.index(originalString.endIndex, offsetBy: -1)
                if originalString[gmtIndicatorIndex] == "Z" {
                    let targetDate = "2014-01-01T00:00:00"
                    let baseDateIndex = originalString.index(originalString.startIndex, offsetBy: targetDate.characters.count)
                    currentString = originalString.substring(to: baseDateIndex)
                    hasMiliseconds = true
                }
            case 25:
                // Copy all the date excluding the timezone also set `hasTimezone` to true.
                // Current date: 2014-01-01T00:00:00+00:00
                // Will become:  2014-01-01T00:00:00
                // Unit test E and F
                let evaluatedDate = "2014-01-01T00:00:00+00"
                let timeZoneColonIndex = originalString.index(originalString.startIndex, offsetBy: evaluatedDate.characters.count)
                if originalString[timeZoneColonIndex] == ":" {
                    let targetDate = "2014-01-01T00:00:00"
                    let baseDateIndex = originalString.index(originalString.startIndex, offsetBy: targetDate.characters.count)
                    currentString = originalString.substring(to: baseDateIndex)
                    hasFullTimeZone = true
                }
            case 27:
                // Copy all the date excluding the microseconds and the Z.
                // Current date: 2015-09-10T00:00:00.184968Z
                // Will become:  2015-09-10T00:00:00
                // Unit test G
                let evaluatedDate = "2015-09-10T00:00:00"
                let timeZoneColonIndex = originalString.index(originalString.startIndex, offsetBy: evaluatedDate.characters.count)
                let gmtIndicatorIndex = originalString.index(originalString.endIndex, offsetBy: -1)
                if originalString[timeZoneColonIndex] == "." && originalString[gmtIndicatorIndex] == "Z" {
                    let targetDate = "2014-01-01T00:00:00"
                    let baseDateIndex = originalString.index(originalString.startIndex, offsetBy: targetDate.characters.count)
                    currentString = originalString.substring(to: baseDateIndex)
                    hasMicroseconds = true
                }
            case 28:
                // Copy all the date excluding the microseconds and the timezone.
                // Current date: 2015-09-10T13:47:21.116+0000
                // Will become:  2015-09-10T13:47:21
                // Unit test H
                let evaluatedDate = "2015-09-10T13:47:21.116"
                let timeZoneColonIndex = originalString.index(originalString.startIndex, offsetBy: evaluatedDate.characters.count)
                if originalString[timeZoneColonIndex] == "+" {
                    let targetDate = "2014-01-01T00:00:00"
                    let baseDateIndex = originalString.index(originalString.startIndex, offsetBy: targetDate.characters.count)
                    currentString = originalString.substring(to: baseDateIndex)
                    hasMiliseconds = true
                    hasNormalizedTimeZone = true
                }
            case 29:
                // Copy all the date excluding the miliseconds and the timezone also set `hasTimezone` to true.
                // Current date: 2015-06-23T12:40:08.000+02:00
                // Will become:  2015-06-23T12:40:08
                // Unit test I
                let evaluatedDate = "2015-06-23T12:40:08.000+02"
                let timeZoneColonIndex = originalString.index(originalString.startIndex, offsetBy: evaluatedDate.characters.count)
                if originalString[timeZoneColonIndex] == ":" {
                    let targetDate = "2015-06-23T12:40:08"
                    let baseDateIndex = originalString.index(originalString.startIndex, offsetBy: targetDate.characters.count)
                    currentString = originalString.substring(to: baseDateIndex)
                    hasFullTimeZone = true
                    hasMiliseconds = true
                }
            case 32:
                // Copy all the date excluding the microseconds and the timezone also set `hasTimezone` to true.
                // Current date: 2015-08-23T09:29:30.007450+00:00
                // Will become:  2015-08-23T09:29:30
                // Unit test J
                let evaluatedDate = "2015-08-23T09:29:30.007450+00"
                let timeZoneColonIndex = originalString.index(originalString.startIndex, offsetBy: evaluatedDate.characters.count)
                if originalString[timeZoneColonIndex] == ":" {
                    let targetDate = "2015-06-23T12:40:08"
                    let baseDateIndex = originalString.index(originalString.startIndex, offsetBy: targetDate.characters.count)
                    currentString = originalString.substring(to: baseDateIndex)
                    hasFullTimeZone = true
                    hasMicroseconds = true
                }
            default:
                fatalError("Your date format \(originalString) is not supported, please file a bug report asking it to be added.")
                break
            }

            // Timezone
            let currentLength = currentString.characters.count
            if hasFullTimeZone {
                // Convert time zone from hours and minutes and append it to the current string.
                // Original date: 2015-06-23T14:40:08.000+02:00
                // Current date:  2015-06-23T14:40:08
                // Will become:   2015-06-23T14:40:08+0200
                var timeZone: String {
                    let endIndex = originalString.index(originalString.endIndex, offsetBy: -"+02:00".characters.count)
                    let fullTimeZone = originalString.substring(from: endIndex)

                    return fullTimeZone.replacingOccurrences(of: ":", with: "")
                }
                currentString = currentString.appending(timeZone)
            } else if hasNormalizedTimeZone {
                // Trim time zone and append it to the current string.
                // Original date: 2015-06-23T14:40:08.000+0200
                // Current date:  2015-06-23T14:40:08
                // Will become:   2015-06-23T14:40:08+0200
                var timeZone: String {
                    let endIndex = originalString.index(originalString.endIndex, offsetBy: -"+0200".characters.count)

                    return originalString.substring(from: endIndex)
                }
                currentString = currentString.appending(timeZone)
            } else {
                // Add GMT timezone to the end of the string
                // Current date: 2015-09-10T00:00:00
                // Will become:  2015-09-10T00:00:00+0000
                currentString = currentString.appending("+0000")
            }

            // Parse the formatted date using `strptime`.
            // %F: Equivalent to %Y-%m-%d, the ISO 8601 date format
            //  T: The date, time separator
            // %T: Equivalent to %H:%M:%S
            // %z: An RFC-822/ISO 8601 standard timezone specification
            var tmc: tm = tm(tm_sec: 0, tm_min: 0, tm_hour: 0, tm_mday: 0, tm_mon: 0, tm_year: 0, tm_wday: 0, tm_yday: 0, tm_isdst: 0, tm_gmtoff: 0, tm_zone: nil)
            if strptime(UnsafeMutablePointer(mutating: currentString), "%FT%T%z", &tmc) == nil {
                return nil
            }

            var time = Double(mktime(&tmc))

            if hasCentiseconds || hasMiliseconds || hasMicroseconds {
                let trimmedDate = dateString.substring(from: "2015-09-10T00:00:00.".endIndex)

                if hasCentiseconds {
                    let centisecondsString = trimmedDate.substring(to: "00".endIndex)
                    if let doubleString = Double(centisecondsString) {
                        let centiseconds = doubleString / 100.0
                        time += centiseconds
                    }
                }

                if hasMiliseconds {
                    let milisecondsString = trimmedDate.substring(to: "000".endIndex)
                    if let doubleString = Double(milisecondsString) {
                        let miliseconds = doubleString / 1000.0
                        time += miliseconds
                    }
                }

                if hasMicroseconds {
                    let microsecondsString = trimmedDate.substring(to: "000000".endIndex)
                    if let doubleString = Double(microsecondsString) {
                        let microseconds = doubleString / 1000000.0
                        time += microseconds
                    }
                }
            }

            self.init(timeIntervalSince1970: time)
        }
    }

    public init(unixTimestampNumber: NSNumber) {
        self.init(unixTimestampString: unixTimestampNumber.stringValue)
    }

    public init(unixTimestampString: String) {
        var parsedString = unixTimestampString
        let validUnixTimestamp = "1441843200"
        let validLength = validUnixTimestamp.characters.count
        if unixTimestampString.characters.count > validLength {
            parsedString = unixTimestampString.substring(to: validUnixTimestamp.endIndex)
        }

        let numberFormatter = NumberFormatter()
        numberFormatter.numberStyle = .decimal
        if let unixTimestampNumber = numberFormatter.number(from: parsedString) {
            self.init(timeIntervalSince1970: unixTimestampNumber.doubleValue)
        } else {
            fatalError("Invalid unix timestamp \(unixTimestampString), please file an issue to get support for your date type")
        }
    }
}

public extension String {
    public func dateType() -> DateType {
        let regex = try! NSRegularExpression(pattern: "-", options: [.caseInsensitive])
        let items = regex.matches(in: self, options: [], range: NSRange(location: 0, length: self.characters.count))
        let ranges = items.map { $0.range }
        if ranges.count > 1 {
            return .iso8601
        } else {
            return .unixTimestamp
        }
    }
}


Sources/NSDate+Parser.m
#import "NSDate+Parser.h"

@implementation NSDate (Parser)

+ (NSDate *)dateFromDateString:(NSString *)dateString {
    NSDate *parsedDate = nil;

    DateType dateType = [dateString dateType];
    switch (dateType) {
        case iso8601: {
            parsedDate = [self dateFromISO8601String:dateString];
        } break;
        case unixTimestamp: {
            parsedDate = [self dateFromUnixTimestampString:dateString];
        } break;
        default: break;
    }

    return parsedDate;
}

+ (NSDate *)dateFromISO8601String:(NSString *)dateString {
    if (!dateString || [dateString isEqual:[NSNull null]]) {
        return nil;
    }

    // Parse string
    else if ([dateString isKindOfClass:[NSString class]]) {
        if ([dateString length] == [HYPPropertyMapperDateNoTimestampFormat length]) {
            NSMutableString *mutableRemoteValue = [dateString mutableCopy];
            [mutableRemoteValue appendString:HYPPropertyMapperTimestamp];
            dateString = [mutableRemoteValue copy];
        }

        const char *originalString = [dateString cStringUsingEncoding:NSUTF8StringEncoding];
        size_t originalLength = strlen(originalString);
        if (originalLength == 0) {
            return nil;
        }

        char currentString[25] = "";
        BOOL hasTimezone = NO;
        BOOL hasCentiseconds = NO;
        BOOL hasMiliseconds = NO;
        BOOL hasMicroseconds = NO;

        // ----
        // In general lines, if a Z is found, then the Z is removed since all dates operate
        // in GMT as a base, unless they have timezone, and Z is the GMT indicator.
        //
        // If +00:00 or any number after + is found, then it means that the date has a timezone.
        // This means that `hasTimezone` will have to be set to YES, and since all timezones go to
        // the end of the date, then they will be parsed at the end of the process and appended back
        // to the parsed date.
        //
        // If after the date theres `.` and a number `2014-03-30T09:13:00.XXX` the `XXX` is the milisecond
        // then `hasMiliseconds` will be set to YES. The same goes for `XX` decisecond (hasCentiseconds set to YES).
        // and microseconds `XXXXXX` (hasMicroseconds set yo YES).
        //
        // If your date format is not supported, then you'll get "Signal Sigabrt". Just ask your format to be included.
        // ----

        // Copy all the date excluding the Z.
        // Current date: 2014-03-30T09:13:00Z
        // Will become:  2014-03-30T09:13:00
        // Unit test H
        if (originalLength == 20 && originalString[originalLength - 1] == 'Z') {
            strncpy(currentString, originalString, originalLength - 1);
        }

        // Copy all the date excluding the timezone also set `hasTimezone` to YES.
        // Current date: 2014-01-01T00:00:00+00:00
        // Will become:  2014-01-01T00:00:00
        // Unit test B and C
        else if (originalLength == 25 && originalString[22] == ':') {
            strncpy(currentString, originalString, 19);
            hasTimezone = YES;
        }

        // Copy all the date excluding the miliseconds and the Z.
        // Current date: 2014-03-30T09:13:00.000Z
        // Will become:  2014-03-30T09:13:00
        // Unit test G
        else if (originalLength == 24 && originalString[originalLength - 1] == 'Z') {
            strncpy(currentString, originalString, 19);
            hasMiliseconds = YES;
        }

        // Copy all the date excluding the miliseconds and the timezone also set `hasTimezone` to YES.
        // Current date: 2015-06-23T12:40:08.000+02:00
        // Will become:  2015-06-23T12:40:08
        // Unit test A
        else if (originalLength == 29 && originalString[26] == ':') {
            strncpy(currentString, originalString, 19);
            hasTimezone = YES;
            hasMiliseconds = YES;
        }

        // Copy all the date excluding the microseconds and the timezone also set `hasTimezone` to YES.
        // Current date: 2015-08-23T09:29:30.007450+00:00
        // Will become:  2015-08-23T09:29:30
        // Unit test D
        else if (originalLength == 32 && originalString[29] == ':') {
            strncpy(currentString, originalString, 19);
            hasTimezone = YES;
            hasMicroseconds = YES;
        }

        // Copy all the date excluding the microseconds and the timezone.
        // Current date: 2015-09-10T13:47:21.116+0000
        // Will become:  2015-09-10T13:47:21
        // Unit test E
        else if (originalLength == 28 && originalString[23] == '+') {
            strncpy(currentString, originalString, 19);
        }

        // Copy all the date excluding the microseconds and the Z.
        // Current date: 2015-09-10T00:00:00.184968Z
        // Will become:  2015-09-10T00:00:00
        // Unit test F
        else if (originalString[19] == '.' && originalString[originalLength - 1] == 'Z') {
            strncpy(currentString, originalString, 19);
        }

        // Copy all the date excluding the miliseconds.
        // Current date: 2016-01-09T00:00:00.00
        // Will become:  2016-01-09T00:00:00
        // Unit test J
        else if (originalLength == 22 && originalString[19] == '.') {
            strncpy(currentString, originalString, 19);
            hasCentiseconds = YES;
        }

        // Poorly formatted timezone
        else {
            strncpy(currentString, originalString, originalLength > 24 ? 24 : originalLength);
        }

        // Timezone
        size_t currentLength = strlen(currentString);
        if (hasTimezone) {
            // Add the first part of the removed timezone to the end of the string.
            // Orignal date: 2015-06-23T14:40:08.000+02:00
            // Current date: 2015-06-23T14:40:08
            // Will become:  2015-06-23T14:40:08+02
            strncpy(currentString + currentLength, originalString + originalLength - 6, 3);

            // Add the second part of the removed timezone to the end of the string.
            // Original date: 2015-06-23T14:40:08.000+02:00
            // Current date:  2015-06-23T14:40:08+02
            // Will become:   2015-06-23T14:40:08+0200
            strncpy(currentString + currentLength + 3, originalString + originalLength - 2, 2);
        } else {
            // Add GMT timezone to the end of the string
            // Current date: 2015-09-10T00:00:00
            // Will become:  2015-09-10T00:00:00+0000
            strncpy(currentString + currentLength, "+0000", 5);
        }

        // Add null terminator
        currentString[sizeof(currentString) - 1] = 0;

        // Parse the formatted date using `strptime`.
        // %F: Equivalent to %Y-%m-%d, the ISO 8601 date format
        //  T: The date, time separator
        // %T: Equivalent to %H:%M:%S
        // %z: An RFC-822/ISO 8601 standard timezone specification
        struct tm tm;
        if (strptime(currentString, "%FT%T%z", &tm) == NULL) {
            return nil;
        }

        time_t timeStruct = mktime(&tm);
        double time = (double)timeStruct;

        if (hasCentiseconds || hasMiliseconds || hasMicroseconds) {
            NSString *trimmedDate = [dateString substringFromIndex:@"2015-09-10T00:00:00.".length];

            if (hasCentiseconds) {
                NSString *centisecondsString = [trimmedDate substringToIndex:@"00".length];
                double centiseconds = centisecondsString.doubleValue / 100.0;
                time += centiseconds;
            }

            if (hasMiliseconds) {
                NSString *milisecondsString = [trimmedDate substringToIndex:@"000".length];
                double miliseconds = milisecondsString.doubleValue / 1000.0;
                time += miliseconds;
            }

            if (hasMicroseconds) {
                NSString *microsecondsString = [trimmedDate substringToIndex:@"000000".length];
                double microseconds = microsecondsString.doubleValue / 1000000.0;
                time += microseconds;
            }
        }

        return [NSDate dateWithTimeIntervalSince1970:time];
    }

    NSAssert1(NO, @"Failed to parse date: %@", dateString);
    return nil;
}

+ (NSDate *)dateFromUnixTimestampNumber:(NSNumber *)unixTimestamp {
    return [self dateFromUnixTimestampString:[unixTimestamp stringValue]];
}

+ (NSDate *)dateFromUnixTimestampString:(NSString *)unixTimestamp {
    NSString *parsedString = unixTimestamp;

    NSString *validUnixTimestamp = @"1441843200";
    NSInteger validLength = [validUnixTimestamp length];
    if ([unixTimestamp length] > validLength) {
        parsedString = [unixTimestamp substringToIndex:validLength];
    }

    NSNumberFormatter *numberFormatter = [NSNumberFormatter new];
    numberFormatter.numberStyle = NSNumberFormatterDecimalStyle;
    NSNumber *unixTimestampNumber = [numberFormatter numberFromString:parsedString];
    NSDate *date = [[NSDate alloc] initWithTimeIntervalSince1970:unixTimestampNumber.doubleValue];

    return date;
}

@end

@implementation NSString (Parser)

- (DateType)dateType {
    if ([self containsString:@"-"]) {
        return iso8601;
    } else {
        return unixTimestamp;
    }
}

@end
